### ISSUE TITLE
egress network policy allows ingress network policy to be bypassed

### ISSUE THREAD

**Issue by shane-kearns** [k3s-io/k3s#2390]
**Environmental Info:**
K3s Version: k3s version v1.17.9+k3s1 (000a361f)

Node(s) CPU architecture, OS, and Version: 

Linux skearns-07 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:

1 master, 3 agents

**Describe the bug:**

When only "ingress" network policies are used, they work correctly.
When ingress and egress network policies are used, a pod "A" selected by an egress policy can contact a pod "B" which is selected by an ingress policy that should forbid this.

This looks similar to https://github.com/cloudnativelabs/kube-router/issues/916 (fixed in kube-router 1.0.0, however k3s is using a forked / vendored version from prior to 0.4.0 based on the comments in pkg/agent/netpol/network_policy_controller.go)

Calico on azure AKS behaves as expected.

**Steps To Reproduce:**

- Installed K3s: "--no-deploy traefik --default-local-storage-path /opt/local-path-provisioner"
- Deploy pods "A", "B", "C" each with a different "app" label
- Deploy an ingress network policy, selecting pod "B", which allows ingress from pod "C" only
- verify C -> B works, A -> B blocked by firewall
- Deploy an egress network policy, selecting pod "A" which permits egress to port 53 (anywhere), or the default namespace

Note: need to wait a minute for iptables to be updated after changes to network policies

**Expected behavior:**

A -> C works, A -> B is blocked by firewall

**Actual behavior:**

A -> B is not blocked

**Additional context / logs:**

**Comment by rancher-max** [comment]
#### Verified using `v1.18.10+k3s1`

See specs below.

**pod.yaml** (replace names and app with "b" and "c" as appropriate)
```
apiVersion: v1
kind: Pod
metadata:
  name: a
  labels:
    app: app-a
spec:
  containers:
    - name: nginx-a
      image: nginx
```

**ingress.yaml**
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: ingress-np
  namespace: default
spec:
  podSelector:
    matchLabels:
      app: app-b
  policyTypes:
  - Ingress
  ingress:
  - from:
    - podSelector:
        matchLabels:
          app: app-c
```

**egress.yaml** (note I used port 80 as that is the port I am curling for nginx)
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: egress-np
  namespace: default
spec:
  podSelector:
    matchLabels:
      app: app-a
  policyTypes:
  - Egress
  egress:
  - to:
    ports:
    - protocol: TCP
      port: 80
```

- I also verified a general "allow-all" ingress with the same results
egress.yaml
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: egress-np
  namespace: default
spec:
  podSelector:
    matchLabels:
      app: app-a
  policyTypes:
  - Egress
  egress:
  - {}
```

**Comment by parente** [comment]
In case it's helpful, I ran into what I think is a variant of this issue on `v1.19.5+k3s2`.

Prohibiting all ingress in namespace A and allowing all egress in namespace B allows pods in B to access pods in A when they are colocated on the same node. I *think* this is a bug in that egress rules in B should not be able to override ingress rules in A, but perhaps I'm mistaken and this is a (very sharp) edge case.

<details>
<summary>Manifests I used to reproduce the issue (click to unfold)</summary>

```
# Create two namespaces
apiVersion: v1
kind: Namespace
metadata:
  name: ns1
  labels:
    name: ns1
---
apiVersion: v1
kind: Namespace
metadata:
  name: ns2
  labels:
    name: ns2
---
# Run a web server on every node in one namespace
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: whoami
  namespace: ns1
  labels:
    app.kubernetes.io/name: whoami
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: whoami
  template:
    metadata:
      labels:
        app.kubernetes.io/name: whoami
    spec:
      containers:
        - name: whoami
          image: traefik/whoami
          ports:
            - name: http
              containerPort: 80
              protocol: TCP
---
# Run a busybox on every node in the other namespace so we can shell
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: busybox
  namespace: ns2
  labels:
    app.kubernetes.io/name: busybox
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: whoami
  template:
    metadata:
      labels:
        app.kubernetes.io/name: whoami
    spec:
      containers:
        - name: busybox
          image: busybox:1.28 # Pinned version, see https://github.com/kubernetes/kubernetes/issues/66924
          command:
            - sleep
            - "3600"
---
# Expose the web server as a normal service
apiVersion: v1
kind: Service
metadata:
  name: whoami
  namespace: ns1
spec:
  type: ClusterIP
  selector:
    app.kubernetes.io/name: whoami
  ports:
    - name: http
      port: 80
      targetPort: 80
---
# Declare a network policy that disallows all ingress to the namespace containing the server
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: disallow-all-ingress
  namespace: ns1
spec:
  podSelector: {}
  policyTypes:
    - Ingress
---
# Declare a network policy that allows all egress from the namespace containing the shell
# BUG: Shells can access the server when its located on the same node as the shell when this egress
# policy is applied. This holds true when access the server pod directly or via the service.
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all-egress
  namespace: ns2
spec:
  podSelector: {}
  policyTypes:
    - Egress
  egress:
    - {}
```
</details>

Steps I took to test connectivity after applying the above:

1. Create a shell in one of the busybox pods
2. `wget http://ip-of-whoami-pod-on-another-node -O-` # good, always fails
3. `wget http://ip-of-whoami-pod-on-same-node-as-this-pod -O-` # bad, always succeeds
4. `wget http://whoami.ns1 -O-` # bad, succeeds when the service happens to route to the colocated pod

**Comment by brandond** [comment]
@parente I believe this is a known issue with network policies? https://kubernetes.io/docs/concepts/services-networking/network-policies/#what-you-can-t-do-with-network-policies-at-least-not-yet
> What you can't do with network policies (at least, not yet)
> * The ability to prevent loopback or incoming host traffic (Pods cannot currently block localhost access, nor do they have the ability to block access from their resident node).

**Comment by parente** [comment]
Thanks for the pointer, @brandond. I took that to meant that traffic from services running on the host node itself could not be affected by policies, not other pods. 

What's interesting is that if I remove the egress policy from namespace B in my example, the ingress rules in A do prevent traffic from pods on the same host. (Maybe by sheer luck given the various ways we can read the parenthetical in the k8s doc.)

**Comment by brandond** [comment]
I will admit that I'm not sure exactly sure what the intended behavior is here. The docs do note that rules are additive, not subtractive - which I would take to mean that an allow on either side is sufficient? You might take this question upstream.

**Comment by brandond** [comment]
If this was indeed caused by cloudnativelabs/kube-router#916, it should have been resolved in #2867. @parente, are you able to retest?

**Comment by shane-kearns** [comment]
Testing of v1.20.8 (which appears to include the mentioned PR's merge commit) was promising.
An egress policy with a pod selector did not interfere with other pods' ingress policies (even with the pod being located on the same node).
However a default egress policy intended to prevent pods making connections outside the cluster still allows ingress rules to be bypassed:
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: default-deny-egress
  namespace: default
spec:
  egress:
  - to:
    - namespaceSelector: {}
  podSelector: {}
  policyTypes:
  - Egress
```
This egress policy does correctly block connections to external services

This same policy works as expected with Calico on Azure - pods in the default namespace cannot connect outside the cluster, and cannot connect to pods inside the cluster where those pods are protected by an ingress policy (unless allowed by the ingress policy)

So this has been substantially improved in that it no longer prevents egress policies being used at all, but has not been fixed yet.

The spec is ambiguous about rules being additive, but the interpretation elsewhere is that the union of all egress rules is intersected with the union of all ingress rules (equivalent to having an ingress firewall rule set and an egress firewall rule set that act independently).
Certainly default rules are much less useful if the ingress and egress rules are combined with union

**Comment by brandond** [comment]
We are about to update kube-router again in #3557, if you're up for testing a commit build. I'm not sure if the behavior you're currently describing has been fixed though, or if it's simply a matter of interpreting the spec differently.

**Comment by fapatel1** [comment]
@shane-kearns would you be able to test this out to see if the issue is still there?

**Comment by shane-kearns** [comment]
This appears to be working correctly in v1.21.4
(so fixed by #3595 )

**Comment by brandond** [comment]
Thanks for the confirmation! Closing, as resolved.
